src/common/browser-adapters/browser-adapter-event-manager.ts

```diff
@@ -51,78 +50,90 @@ export class BrowserAdapterEventManager {
+            } else {
+                stillDeferred.push(deferredEvent);
+            }
+        });
+        this.deferredEvents = stillDeferred;
     }
 
-    private async forwardEventToApplicationListener(
+    private forwardEventToApplicationListener(
         listener: ApplicationListener,
         eventArgs: any[],
     ): Promise<any> {
-        const result = listener(...eventArgs);
+        let result: any;
+        try {
+            result = listener(...eventArgs);
+        } catch (error) {
+            console.error('Error thrown in application listener: ', error);
+            throw error;
+        }
         if (!!result && typeof result.then === 'function') {
             //wrapping application listener promise responses in a 4-minute promise race
             //prevents the service worker going idle before a response is sent
-            await this.promiseFactory.timeout(result, FOUR_MINUTES);
+            return this.promiseFactory.timeout(result, FOUR_MINUTES);
         } else {
             // it is possible that this is the result of a fire and forget listener
             // wrap in 2-minute timeout to ensure it completes during service worker lifetime
-            return this.timeoutFactory.timeoutType === TimeoutType.Alarm
-                ? this.timeoutFactory.createTimeout(
-                      () => Promise.resolve(result),
-                      TWO_MINUTES,
-                      `wrapper-timeout-${Date.now()}`,
-                  )
-                : this.timeoutFactory.createTimeout(() => Promise.resolve(result), TWO_MINUTES);
+            globalThis.setTimeout(() => {
+                Promise.resolve(result);
+            }, TWO_MINUTES);
+            return Promise.resolve(result);
         }
-        return result;
     }
 
-    public processEvent(eventType: string, eventArgs: any[], isDeferred?: boolean): boolean {
-        if (this.eventsToApplicationListenersMapping[eventType]) {
-            this.eventsToApplicationListenersMapping[eventType].forEach(applicationListener => {
-                this.forwardEventToApplicationListener(applicationListener, eventArgs);
-            });
-            return true;
+    private deferEvent(eventDetails: EventDetails): Promise<any> {
+        const deferredResolution = this.promiseFactory.externalResolutionPromise();
+
+        this.deferredEvents.push({
+            deferredResolution: deferredResolution,
+            ...eventDetails,
+        });
+
+        return deferredResolution.promise;
+    }
+
+    public processEvent(eventType: string, eventArgs: any[]): Promise<any> {
+        const applicationListener = this.eventsToApplicationListenersMapping[eventType];
+        if (applicationListener) {
+            return this.forwardEventToApplicationListener(applicationListener, eventArgs);
         } else {
-            if (!isDeferred) {
-                const eventDetails: EventDetails = {
+            const alreadyDeferred = this.deferredEvents.find(
```

Comment:
This `find` call gives `processDeferedEvents` an `O(deferredEvents.length ^ 2)` time complexity, and we do see thousands of events in normal usage of the app. "Are we in an already-deferred" path is already known in each of the 2 `processEvent` callsites, and one of them doesn't actually use the return value of `processEvent`; maybe instead of the current contents of this else branch, just return `null` and update L32 to be something along the lines of `return processEvent(e, ...args) ?? deferEvent(e, ...args);`?